include/deal.II/multigrid/mg_transfer_interface.h

```diff
@@ -0,0 +1,424 @@
+   * via weights.
+   */
+  bool fine_element_is_continuous;
+
+  /**
+   * Weights for continuous elements.
+   */
+  std::vector<Number> weights;
+
+  /**
+   * 1D prolongation matrix.
+   */
+  AlignedVector<VectorizedArray<Number>> prolongation_matrix_1d;
+
+  /**
+   * DoF indices of the coarse cells.
+   */
+  std::vector<unsigned int> level_dof_indices_coarse;
+
+  /**
+   * DoF indices of the children of the coarse cells.
+   */
+  std::vector<unsigned int> level_dof_indices_fine;
+
+  /**
+   * Print internal data structures to stream @p out.
+   */
+  template <typename Stream>
+  void
+  print(Stream &out) const;
+};
+
+
+
+/**
+ * Class for transfer between two multigrid levels.
+ */
+template <int dim, typename Number>
+class Transfer
+{
+public:
+  /**
+   * Print internal data structures to stream @p out.
+   */
+  template <typename Stream>
+  void
+  print_internal(Stream &out) const;
+
+  /**
+   * Perform prolongation.
+   */
+  void
+  prolongate(const unsigned int,
+             LinearAlgebra::distributed::Vector<Number> &      dst,
+             const LinearAlgebra::distributed::Vector<Number> &src) const;
+
+  /**
+   * Perform restriction.
+   */
+  void
+  restrict_and_add(const unsigned int,
+                   LinearAlgebra::distributed::Vector<Number> &      dst,
+                   const LinearAlgebra::distributed::Vector<Number> &src) const;
+
+private:
+  /**
+   * Transfer schemes.
+   */
+  std::vector<TransferScheme<Number>> schemes;
+
+  /**
+   * Partitioner needed by the intermediate vector.
+   */
+  std::shared_ptr<const Utilities::MPI::Partitioner> partitioner_fine;
+
+  /**
+   * TODO: needed?
+   */
+  std::shared_ptr<const Utilities::MPI::Partitioner> partitioner_coarse;
+
+  /**
+   * Internal vector needed for collecting all degrees of freedom of the
+   * children cells.
+   */
+  mutable LinearAlgebra::distributed::Vector<Number> vec_fine;
+
+  /**
+   * TODO: needed?
+   */
+  mutable LinearAlgebra::distributed::Vector<Number> vec_coarse;
+
+  /**
+   * Constraint matrix on coarse level.
+   */
+  AffineConstraints<Number> constraint_coarse;
+
+  friend class MGTransferUtil::Implementation;
+};
+
+
+
+/**
+ * Implementation of the MGTransferBase interface for which the transfer
+ * operations is implemented in a matrix-free way based on the interpolation
+ * matrices of the underlying finite element. This requires considerably less
+ * memory than MGTransferPrebuilt and can also be considerably faster than that
+ * variant. In contrast to MGTransferMatrixFree, the user can provide separate
+ * transfer operators between each level.
+ *
+ * This class currently only works for tensor-product finite elements based on
+ * FE_Q and FE_DGQ elements. Systems involving multiple components of
+ * one of these element, as well as, systems with different elements or other
+ * elements are currently not implemented.
+ */
+template <typename MatrixType>
+class MGTransferMatrixFreeNew
```

Comment:
The naming is not ideal. Can it be a replacement for the existing MGTransferMatrixFree? We can think about backwards-compatibility separately.

Reply:
As it is now, it is not a replacement yet. I have not incorporated the interface matrices since I did not need them for global coarsening and in the hp context. 

Nevertheless, merging both would be nice. For example the following code snippet:

https://github.com/dealii/dealii/blob/9e97cd69e4d72012f9d913c8f0cae69e8f8042a4/source/multigrid/mg_transfer_matrix_free.cc#L373-L508

is more or less identically in my code (with the difference that different local prolongation matrices can be used depending on the cells and that cells might have children or not).

Reply:
maybe MGTransferGlobalCoarsening?